Extract filename obfuscation into module

alyssais opened the pull request:
Here's how I would extract the filename obfuscation logic into a re-usable module. I'll do a review after I've created the PR and go through and explain it a bit. :)

**Beware**. It was taking a very long time to get a local instance of Mastodon working, and I decided it wasn't worth the effort, so this is untested, but it should be pretty easy to verify it works the same if you're able to run it.

alyssais committed:
Extract filename obfuscation into module

alyssais commented on app/models/concerns/obfuscate_filename.rb:
```diff
@@ -0,0 +1,14 @@
+module ObfuscateFilename
```
I pulled this out into its own module. Alternatively, since it's only really one method, it could have gone into `ApplicationController`. I'm a fan of many small modules and a clean `ApplicationController` personally, but others disagree.

alyssais replied:
Every file in the `app/models/concerns` directory is automatically loaded by Rails, so there's no need to import it in other files.

alyssais commented on app/models/concerns/obfuscate_filename.rb:
```diff
@@ -0,0 +1,14 @@
+module ObfuscateFilename
+  extend ActiveSupport::Concern
```
`ActiveSupport::Concern` is a module that Rails provides that just makes Ruby modules a bit nicer to work with. Key things it does:

- You don't have to worry about which modules are supposed to be included with `include` and which with `extend` (`include` adds the module methods as instance methods, `extend` makes them class methods). With an `ActiveSupport::Concern`, you can always use `include`.
- An `ActiveSupport::Concern` can declare both instance and class methods, by putting the class methods in an inner `ClassMethods` module. Normal Ruby modules just have to declare a list of methods, and it's up to the includer to make them either all class methods, or all instance methods. This doesn't often end well.
- `ActiveSupport::Concern`s can call `included` and pass a block containing code that will be evaluated when the module is included. This is useful, for example, if you want to declare a `before_action` in a controller. You can do this with normal Ruby modules, but `ActiveSupport::Concern`'s way is slightly higher-level.

hikari-no-yume commented:
Protip: don't use Docker. Use the docker file as a guide for which commands you need to install the dependencies (there's some npm and you need the Ruby stuff). Then run it locally, again using the docker config files as a guide.

alyssais commented on app/models/concerns/obfuscate_filename.rb:
```diff
@@ -0,0 +1,14 @@
+
+  module ClassMethods
+    def self.obfuscate_filename(*args)
+      before_action { obfuscate_filename(*args) }
```
This method just takes all its arguments, and declares a `before_action` that just calls an instance method of the same name with all of those arguments. A `before_action` takes a block of code that is run in a controller instance before the main controller method (e.g. `update`, `show`, etc.) is called. The Rails documentation [explains](http://guides.rubyonrails.org/action_controller_overview.html#filters) the order of Rails controller operations in more detail. So, this lets us modify the parameters before the main controller methods get to see them.

alyssais commented on app/models/concerns/obfuscate_filename.rb:
```diff
@@ -0,0 +1,14 @@
+  end
+
+  def obfuscate_filename(path)
+    file = Array(path).reduce { |params, key| params[key] }
```
If `Array` is passed a non-collection object, it will return that object wrapped in an array. If it gets a collection object, it will convert that collection to an array.

In this situation, it allows us to accept both a single symbol, or an array of symbols, and treat them both the same (as an array). This means that the simple case of the file being one level deep in the hash is more elegant.

Rather than taking a single argument here, we could have taken a variable argument where each argument was an element of the path. However, it would have been easy to confuse that with obfuscating multiple filenames, so I didn't do that.

Instead, this could be extended to take variable arguments, and just loop through the body of the method for each argument. That would allow callers to specify multiple files in a single call, which might help keep their controllers tidy.

alyssais commented:
> Protip: don't use Docker.

I did exactly this. I can get the application to start, but I wanted to get all the tests passing (I hate manually testing Rails apps when I could do it way more efficiently with automated tests), and there was some complication with Neo4j using a weird port and the tests expecting to be run from Ngrok(???) so then I gave up.

hikari-no-yume commented:
Oh, you don't even have to worry about Neo4j, that's optional. The production site currently isn't running it, even.

alyssais committed:
Replace a scary Array#reduce with Hash#dig

alyssais commented:
The tests need it, though.

hikari-no-yume commented:
Ah :/

alyssais commented:
Just updated because I remembered Ruby recently added a method for doing exactly what I needed that `reduce` to do, so the one complicated line is now way simpler.

That does mean GitHub has collapsed my (mostly still relevant) comments on it though, so make sure to expand them again.

alyssais committed:
Remove unnecessary Array call

hikari-no-yume commented:
ActionController::RoutingError - undefined method `obfuscate_filename' for Api::V1::MediaController:Class

Hrm.

alyssais commented on app/models/concerns/obfuscate_filename.rb:
```diff
@@ -0,0 +1,14 @@
+  end
+
+  def obfuscate_filename(path)
+    file = params.dig(*path)
```
Okay, to replace my previous comment now that this line is a lot simpler:

Here we use the `*` operator to expand `path` to an argument list if it's an array. This gives callers the option of passing an array or a single element as the `path` parameter. See my [comment](https://github.com/TazeTSchnitzel/mastodon/pull/1#discussion_r89416929) on the previous version of this line for an explanation of why.

alyssais commented:
> Hrm.

Oh…

I forgot to include the module, didn't I?

hikari-no-yume commented:
Yep…

alyssais committed:
Include ObfuscateFilename where it's used

hikari-no-yume commented:
Error persists, I don't know why.

alyssais commented:
Stop your Rails server, run `spring stop` to get Rails to load new files (it's running in the background, you see) then start again.

hikari-no-yume commented:
I had restarted the server. Your code only works when I use `extend` instead of `include`. Well, actually it's broken, but now for a different reason.

alyssais commented:
Ugh, fine, I'll get my instance working.

alyssais committed:
Don't make obfuscate_filename static

alyssais commented:
Oh, it was because I'd declared `ObfuscateFilename::ClassMethods::obfuscate_filename` as a static method on the module, rather than as a module method. How embarrassing.

Changing `module ClassMethods` to `class_methods do` was not part of the fix, but I noticed when I looked at the Rails docs that they've added the `class_methods do` syntax since I last used it and I decided I liked it better because with `module ClassMethods` there's no error if you typo it.

alyssais commented:
When you were extending the module, that was pulling the instance method `obfuscate_filename` up to the class level (where it isn't supposed to be!). So that was why you could call the method, but it wouldn't actually work — there were no `params` for it to modify!

hikari-no-yume commented:
Right, that makes sense.

hikari-no-yume commented:
Hmm, your code works for attachments to posts, but now the settings page is completely broken. Investigating…

alyssais commented:
This is what happens when I can't write unit tests

alyssais commented:
Oh, I know what it is.

alyssais committed:
Check if file parameter is present before changing

This was causing the settings page not to load, because `file` is only
set when updating, not show/edit-ing.

hikari-no-yume commented:
Fantastic. Now, time for some merging magic…

alyssais commented:
Yeah this is urgently in need of a squash!

hikari-no-yume commented:
Merged in https://github.com/Gargron/mastodon/pull/242/commits/cefef2c57197a06d403b9cbe13e969e08aaf5301. Thank you!
